docs/man/man1/hocr.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH HOCR "1" "Jan 2011" "hebocr - Hebrew OCR utility" "User Commands"
.SH NAME
hebocr \- manual page for hocr - Hebrew OCR utility
.SH DESCRIPTION
.SS "Usage:"
.IP
hocr [OPTION...] \- Hebrew OCR utility
.SS "Help Options:"
.TP
\-?, \fB\-\-help\fR
Show help options
.TP
\fB\-\-help\-all\fR
Show all help options
.TP
\fB\-\-help\-file\fR
Show file options
.TP
\fB\-\-help\-image\-proccesing\fR
Show image proccesing options
.TP
\fB\-\-help\-segmentation\fR
Show segmentation options
.TP
\fB\-\-help\-debug\fR
Show debug options
.PP
File options
.TP
\fB\-O\fR, \fB\-\-images\-out\-path\fR=\fIPATH\fR
use PATH for output images
.TP
\fB\-u\fR, \fB\-\-data\-out\fR=\fIFILE\fR
use FILE as output data file name
.TP
\fB\-C\fR, \fB\-\-save\-copy\fR
save a compy of original image
.TP
\fB\-b\fR, \fB\-\-save\-bw\fR
save proccesd bw image
.TP
\fB\-B\fR, \fB\-\-save\-bw\-exit\fR
save proccesd bw image and exit
.TP
\fB\-l\fR, \fB\-\-save\-layout\fR
save layout image
.TP
\fB\-L\fR, \fB\-\-save\-layout\-exit\fR
save layout image and exit
.TP
\fB\-f\fR, \fB\-\-save\-fonts\fR
save fonts
.TP
\fB\-F\fR, \fB\-\-save\-fonts\-exit\fR
save fonts images and exit
.PP
Image proccesing options
.TP
\fB\-T\fR, \fB\-\-thresholding\-type\fR=\fINUM\fR
thresholding type, 0 normal, 1 none, 2 fine
.TP
\fB\-t\fR, \fB\-\-threshold\fR=\fINUM\fR
use NUM as threshold value, 1..100
.TP
\fB\-a\fR, \fB\-\-adaptive\-threshold\fR=\fINUM\fR
use NUM as adaptive threshold value, 1..100
.TP
\fB\-s\fR, \fB\-\-scale\fR=\fISCALE\fR
scale input image by SCALE 1..9, 0 auto
.TP
\fB\-S\fR, \fB\-\-no\-auto\-scale\fR
do not auto acale image
.TP
\fB\-q\fR, \fB\-\-rotate\fR=\fIDEG\fR
rotate image clockwise in deg.
.TP
\fB\-Q\fR, \fB\-\-no\-auto\-rotate\fR
do not auto rotate image
.PP
Segmentation options
.TP
\fB\-c\fR, \fB\-\-colums\fR setup=NUM
colums setup: 1.. #colums, 0 auto, 255 free
.TP
\fB\-x\fR, \fB\-\-slicing\fR=\fINUM\fR
use NUM as font slicing threshold, 1..250
.TP
\fB\-X\fR, \fB\-\-slicing\-width\fR=\fINUM\fR
use NUM as font slicing width, 50..250
.TP
\fB\-w\fR, \fB\-\-font\-spacing\fR=\fINUM\fR
font spacing: tight ..\-1, 0, 1.. spaced
.PP
Debug options
.TP
\fB\-g\fR, \fB\-\-draw\-grid\fR
draw grid on output images
.TP
\fB\-d\fR, \fB\-\-debug\fR
print debuging information while running
.TP
\fB\-D\fR, \fB\-\-debug\-extra\fR
print extra debuging information
.TP
\fB\-y\fR, \fB\-\-font\-filter\fR=\fINUM\fR
debug a font filter, use filter NUM
.TP
\fB\-Y\fR, \fB\-\-font\-filter\-list\fR
print a list of debug a font filters
.TP
\fB\-j\fR, \fB\-\-font\-num\-out\fR
print font numbers in output text
.SS "Application Options:"
.TP
\fB\-i\fR, \fB\-\-image\-in\fR=\fIFILE\fR
use FILE as input image file name
.TP
\fB\-o\fR, \fB\-\-text\-out\fR=\fIFILE\fR
use FILE as output text file name
.TP
\fB\-h\fR, \fB\-\-html\-out\fR
output text in html format
.TP
\fB\-N\fR, \fB\-\-no\-gtk\fR
do not use gtk for file input and output
.TP
\fB\-z\fR, \fB\-\-font\fR=\fINUM\fR
use font NUM
.TP
\fB\-n\fR, \fB\-\-no\-nikud\fR
do not recognize nikud
.TP
\fB\-v\fR, \fB\-\-version\fR
print version information and exit
.PP
libhebocr\-0.11
http://code.google.com/p/hebocr/

Copyright (C) 2005\-2008 Yaacov Zamir <kzamir@walla.co.il>
Copyright (C) 20011 Diego Iastrubni <diegoiast@gmail.com>

.PP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
.TP
.SH "SEE ALSO"
gocr(1), ocrad(1), unpaper(1)


ChangeLog
libhocr (0.10.17)

 * Font recognition fine tuning
 * Add tiff and jpeg formats output to hocr utility
 * Add left to right options to hocr and hocr-gtk utilities
 * Oron Peled <oron_at_actcom.co.il> :
   - Fix hocr-gtk and sane-pygtk desktop files 
   - Fix pkg-config files
   - code cleanups: missings includes and unused vars
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Monday, 27 July 2008 09:04

libhocr (0.10.16)

 * Update hocr-gtk Hebrew translation
 * Add dilation and erosion to hocrpp.h
 * Re Fix python install scripts
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Monday, 22 July 2008 09:04
 
libhocr (0.10.13)

 * Oron Peled <oron_at_actcom.co.il> :
   - Fix errors in examples/hocr-gtk/sane-pygtk.desktop
   - Fix naming in examples/bindings/Makefile.am
   - Use AM_PATH_PYTHON in configure.ac
 * Add more options to hocr-gtk
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Monday, 21 July 2008 09:04
 
libhocr (0.10.12)

 * Fix font segmentation when fonts overlap
 * Re change defaults for image prossecing
 * Fix font slicing bug
 * Sync hocr-gtk and hocr version numbers with libhocr version number
 * Fix memory leaks in hocr-gtk
 * Fix gtk pixbuf save extention problem

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Thursday, 06 July 2008 09:04
 
libhocr (0.10.11)

 * Change default values in hocr command line utility
 * Change default values in hocr-gtk graphical utility
 * Add scale and threshold settings to OO c++ and python bindings

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Thursday, 26 June 2008 09:04
 
libhocr (0.10.10)

 * Fix hocr-gtk example remember window size
 * Fix hocr-gtk example show image filename in title
 * hocr-gtk example requires python bindings 
 * add an ho_array object for arithmetic based operations
 * add an ho_array_hist helper object for ho_array
 * add an ho_array_stat helper object for ho_array's statistics
 * add ah ho_bitmap_hist helper object for ho_bitmap
 * Fix libhocrgtk build on none gtk systems (Bug report: Nadav Mavor)
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Sunday, 10 May 2008 09:04
 
libhocr (0.10.9)

 * Re-Fix auto scaling bug

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Thursday, 8 May 2008 09:04
 
libhocr (0.10.8)

 * Fix auto scaling bug
 * Refine david font 

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Thursday, 8 May 2008 09:04
 
libhocr (0.10.7)

 * hocr-gtk save session parameters
 * remove .py from hocr-gtk example python scripts file names
 * use libtiff for black and white 
   hocr-pixbuf export and import using tiff format
 * hocr example can read black and white tiff image files
   without using gtk
 * use sane-pygtk as the scanner utility
 * make the configure script more flexible

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Thursday, 24 Apr 2008 09:04
 
libhocr (0.10.6)

 * hocr-gtk example less path dependent (use sys.prefix)
 * hocr-gtk example can use tesseract-ocr if present
 * hocr-gtk example can use python-imaging-sane if present to scan images
   from a scanner
 * use libtiff for hocr-bitmap export using tiff format
 * python-hocr does not include hocr-gtk

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Wednesday, 22 Apr 2008 21:04
 
libhocr (0.10.5)

 * fine tune font recognition
 * add wrapper functions
 * add c++ wrapper class
 * add hocr-gtk example (using glade, pygtk and hocr-python)

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Saturday, 26 Jan 2008 21:04

libhocr (0.10.4)

 * change licence to GPL3
 * fine tune options for segmentation functions
 * fine tune reading order functions
 * remove web interface
 * fix pbm load bug (patch: Mikael Ylikoski <ordbrand@gmail.com>)
 * add rotation functions
 * re-add examples (hocr example requers gtk)
 * re-add python bindings (now requires gtk)

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Monday, 14 May 2007 18:12
 
libhocr (0.10.3)

 * font segmentation functions
 * font recognition functions
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Monday, 4 March 2007 12:02

libhocr (0.10.2)

 * paragraph segmentation functions
 * line segmentation functions
 * word segmentation functions
 * reading order sorting functions
 * add php web interface 
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Thursday, 15 February 2007 19:12
 
libhocr (0.10.1)

 * internal image loader can load pnm P6, P5, P4 images
 * internal image saver can save pnm P6, P5 images
 * new functions for color and gray scale images 
 * gtk related functions move to libhocrgtk

 -- Yaacov Zamir <kzamir_at_walla.co.il>  Tuesday, 16 Jan 2007 20:10

libhocr (0.10.0)

 * new API
 * a code cleanup (re-write of most of the code):
    re-write of object/struct definitions.
    re-write of the segmentation and recognition functions.
    re-write of the input-output functions.
    file re-organization and naming.
 * temporarily remove bindings and examples.
 * hocr command line utility requires gtk for input and output
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Sun, 25 Jun 2006 22:13

hocr (0.8.2)

 * change g_file_set_contents to simple fprintf call in file 
   examples/hocr-gtk/callbacks.c
 * add hspell backend to hocr_textbuffer
 * add glib's Character Set Conversion support to hocr_textbuffer
 * add glib's File save support to hocr_textbuffer
 * add gtk's open/save picture files to hocr_pixbuf.
 * default brightness is 128 now
 * new and preetier logo by Shlomi
 * Daniel Nylander: swedish Swedish translation of hocr-gtk
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Sun, 15 Jan 2006 22:13

hocr (0.8.0)

 * Shlomi Israel: hocr logo.
 * Tal. (tal.bav_at_gmail.com): Fix memory leak when unrefing the text structure 
 
 -- Yaacov Zamir <kzamir_at_walla.co.il>  Sat, 14 Jan 2006 22:13

hocr (0.7.2)

 * command line example implements pipes
 * Tal. (tal.bav_at_gmail.com): add save ppm to command line example.
 * move tal's save as pnm to the internal hocr_pixbuf functions.
 * add command line args to hocr-gtk example
 * add save picture to hocr-gtk example
 * add xsane option to hocr-gtk example
 * clean picture from artefacts before ocr it

src/hocr.c

/***************************************************************************
 *            hocr.c
 *
 *  Fri Aug 12 20:13:33 2005
 *  Copyright  2005-2008  Yaacov Zamir
 *  <kzamir@walla.co.il>
 ****************************************************************************/

/*  
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <math.h>

#include "hebocr_globals.h"
#include "hocr.h"


/**
 convert a gray pixbuf to bitmap 

 @param pix_in the input ho_pixbuf
 @param scale the scale to use
 @param no_auto_scale do not try to autoscale
 @param rotate the rotation angle to use
 @param no_auto_rotate do not try to autorotate
 @param adaptive what type of thresholding to use
 @param threshold the threshold to use 0..100
 @param a_threshold the threshold to use for adaptive thresholding 0..100
 @param progress a progress indicator 0..100
 @return newly allocated gray ho_bitmap
 */
ho_bitmap *hocr_image_processing (const ho_pixbuf * pix_in, HEBOCR_IMAGE_OPTIONS *options, int *progress)
{

  ho_bitmap *bitmap_out = NULL;
  ho_bitmap *bitmap_temp = NULL;
  double angle = 0.0;
  int scale_by = 0;
  unsigned char size = 0;

  /* init progress */
  *progress = 0;

  /* get the raw b/w bitmap from the pixbuf */
  bitmap_temp = ho_pixbuf_to_bitmap_wrapper (pix_in,
    options->scale, options->adaptive, options->threshold, options->a_threshold, size);
  if (!bitmap_temp)
    return NULL;

  /* update progress */
  *progress = 25;

  /* do we want to auto scale ? */
  if (!options->scale && !options->no_auto_scale)
  {
    /* get fonts size for autoscale */
    if (ho_dimentions_font_width_height_nikud (bitmap_temp, 6, 200, 6, 200))
      return NULL;

    /* if fonts are too small, re-scale image */
    if (bitmap_temp->font_height < 15)
      scale_by = 3;
    else if (bitmap_temp->font_height < 30)
      scale_by = 2;
    else
      scale_by = 1;

    if (scale_by > 1)
    {
      /* re-create bitmap */
      ho_bitmap_free (bitmap_temp);
      bitmap_temp = ho_pixbuf_to_bitmap_wrapper (pix_in,
	scale_by, options->adaptive, options->threshold, options->a_threshold, size);
      if (!bitmap_temp)
        return NULL;
    }
  }

  /* update progress */
  *progress = 50;

  /* remove very small and very large things */
  bitmap_out =
    ho_bitmap_filter_by_size (bitmap_temp, 3, 3 * bitmap_temp->height / 4, 3,
    3 * bitmap_temp->width / 4);
  ho_bitmap_free (bitmap_temp);
  if (!bitmap_out)
    return NULL;

  /* update progress */
  *progress = 75;

  /* rotate image */
  if (options->rotate)
  {
    bitmap_temp = ho_bitmap_rotate (bitmap_out, options->rotate);
    ho_bitmap_free (bitmap_out);
    if (!bitmap_temp)
      return NULL;

    bitmap_out = bitmap_temp;
  }
  else if (!options->no_auto_rotate)
  {
    /* get fonts size for auto angle */
    if (ho_dimentions_font_width_height_nikud (bitmap_out, 6, 200, 6, 200))
      return NULL;

    angle = ho_dimentions_get_lines_angle (bitmap_out);
    if (angle)
    {
      bitmap_temp = ho_bitmap_rotate (bitmap_out, angle);
      ho_bitmap_free (bitmap_out);
      if (!bitmap_temp)
        return NULL;

      bitmap_out = bitmap_temp;
    }
  }

  return bitmap_out;
}

/**
 new ho_layout 

 @param m_in a pointer to a text bitmap
 @param font_spacing_code -3 tight .. 0 .. 3 spaced
 @param paragraph_setup free text blocks or boxed in columns
 @param slicing_threshold percent of line fill to cut fonts
 @param slicing_width what is a wide font
 @param dir true-ltr false-rtl
 @param progress a progress indicator 0..100
 @return a newly allocated and filled layout
 */
ho_layout *
hocr_layout_analysis (const ho_bitmap * m_in, const int font_spacing_code,
  const int paragraph_setup, const int slicing_threshold,
  const int slicing_width, const int line_leeway, const unsigned char dir_ltr, int *progress)
{
  int cols = paragraph_setup;

  int block_index;

  int line_index;

  int word_index;

  ho_layout *layout_out = NULL;

  /* init progress */
  *progress = 0;

  /* check paragraph_setup */
  if (!cols)
    cols = ho_dimentions_get_columns (m_in);

  /* create a new layout */
  layout_out = ho_layout_new (m_in, font_spacing_code, cols, dir_ltr);
  if (!layout_out)
    return NULL;

  *progress = 10;
  ho_layout_create_block_mask (layout_out);

  /* look for lines inside blocks */
  for (block_index = 0; block_index < layout_out->n_blocks; block_index++)
  {
    ho_layout_create_line_mask (layout_out, block_index);

    /* look for words inside line */
    for (line_index = 0; line_index < layout_out->n_lines[block_index];
      line_index++)
    {
      ho_layout_create_word_mask (layout_out, block_index, line_index);

      /* look for fonts inside word */
      for (word_index = 0;
        word_index < layout_out->n_words[block_index][line_index]; word_index++)
      {
        ho_layout_create_font_mask (layout_out, block_index, line_index,
          word_index, slicing_threshold, slicing_width, line_leeway);
      }

      /* update progress */
      *progress = 100 *
        ((double) block_index /
        (double) layout_out->n_blocks +
        (double) line_index /
        (double) (layout_out->n_lines[block_index] * layout_out->n_blocks));
    }
  }

  return layout_out;
}

/**
 fill a text buffer with fonts recognized from a page layout

 @param l_page the page layout to recognize
 @param s_text_out the text buffer to fill
 @param html output format is html
 @param font_code code for the font to use
 @param nikud recognize nikud
 @param progress a progress indicator 0..100
 @return FALSE
 */
int
hocr_font_recognition (const ho_layout * l_page, ho_string * s_text_out,
  const unsigned char html, int font_code, const unsigned char nikud,
  unsigned char do_linguistics, int *progress)
{
  int block_index;

  int line_index;

  int word_index;

  int font_index;

  int current_font_number = 0;

  int number_of_fonts = l_page->number_of_fonts;

  ho_bitmap *m_text = NULL;

  ho_bitmap *m_mask = NULL;

  ho_bitmap *m_font_main_sign = NULL;

  ho_bitmap *m_font_nikud = NULL;

  char text_out[200];

  const char *font;

  const char *font_nikud;

  const char *font_dagesh;

  const char *font_shin;

  /* init progress */
  *progress = 0;

  /* did we get a text buffer and a layout ? */
  if (!s_text_out || !l_page)
    return TRUE;

  /* loop over the layout */
  for (block_index = 0; block_index < l_page->n_blocks; block_index++)
  {
    /* start of paragraph */
    if (html)
    {
      /* FIXME: text overflow ?! */
      sprintf (text_out,
        "    <div class=\"ocr_par\" id=\"par_%d\" title=\"bbox %d %d %d %d\">\n",
        block_index + 1, l_page->m_blocks_text[block_index]->x,
        l_page->m_blocks_text[block_index]->y,
        l_page->m_blocks_text[block_index]->x +
        l_page->m_blocks_text[block_index]->width,
        l_page->m_blocks_text[block_index]->y +
        l_page->m_blocks_text[block_index]->height);
      ho_string_cat (s_text_out, text_out);
    }

    for (line_index = 0; line_index < l_page->n_lines[block_index];
      line_index++)
    {
      /* start of line */

      /* loop on all the words in this line */
      for (word_index = 0;
        word_index < l_page->n_words[block_index][line_index]; word_index++)
      {
        int word_length = l_page->n_fonts[block_index][line_index][word_index];

        unsigned char word_end = FALSE;

        unsigned char word_start = TRUE;

        int last_char_i = 0;

        int char_i = 0;

        /* start of word */
        for (font_index = 0; font_index < word_length; font_index++)
        {
          word_end = (font_index == (word_length - 1));
          word_start = (font_index == 0);

          /* get font images */

          /* get the font */
          m_text =
            ho_layout_get_font_text (l_page, block_index,
            line_index, word_index, font_index);
          if (!m_text)
            return TRUE;

          /* get font line mask */
          m_mask =
            ho_layout_get_font_line_mask (l_page, block_index,
            line_index, word_index, font_index);
          if (!m_mask)
            return TRUE;

          /* get font main sign */
          m_font_main_sign = ho_font_main_sign (m_text, m_mask);
          if (!m_font_main_sign)
            return TRUE;

          /* recognize font from images */
          last_char_i = char_i;
          font =
            ho_recognize_font (m_font_main_sign, m_mask,
            font_code, do_linguistics, word_end, word_start, &char_i,
            last_char_i);

          /* insert font to text out */
          ho_string_cat (s_text_out, font);

          /* get font nikud */
          if (nikud)
          {
            m_font_nikud = ho_bitmap_clone (m_text);
            if (!m_font_nikud)
              return TRUE;
            ho_bitmap_andnot (m_font_nikud, m_font_main_sign);

            /* recognize font from images */
            font_nikud = ho_recognize_nikud (m_font_nikud, m_mask,
              font_code, &font_dagesh, &font_shin);

            /* free bitmaps */
            ho_bitmap_free (m_font_nikud);
            m_font_nikud = NULL;

            /* insert font nikud to text out */
            ho_string_cat (s_text_out, font_shin);
            ho_string_cat (s_text_out, font_dagesh);
            ho_string_cat (s_text_out, font_nikud);
          }

          /* free bitmaps */
          ho_bitmap_free (m_font_main_sign);
          ho_bitmap_free (m_text);
          ho_bitmap_free (m_mask);

          /* this are empty pointers */
          m_text = m_mask = m_font_main_sign = NULL;

          /* update progress */
          current_font_number++;
          *progress = 100 * current_font_number / number_of_fonts;
        }

        /* end of word */
        ho_string_cat (s_text_out, " ");
      }

      /* end of line */
      if (html)
        ho_string_cat (s_text_out, "<br/>\n");
      else
        ho_string_cat (s_text_out, "\n");
    }

    /* end of block */
    if (html)
      ho_string_cat (s_text_out, "<br/>\n    </div>\n");
    else
      ho_string_cat (s_text_out, "\n");
  }

  return FALSE;
}

int
hocr_do_ocr_fine (const ho_pixbuf * pix_in,
  ho_string * s_text_out,
  HEBOCR_IMAGE_OPTIONS *options,
  const int font_spacing_code, const int paragraph_setup,
  const int slicing_threshold, const int slicing_width, const int line_leeway,
  const unsigned char dir_ltr,
  const unsigned char html, int font_code, const unsigned char nikud,
  const unsigned char do_linguistics, int *progress)
{
  ho_bitmap *m_in = NULL;

  ho_layout *l_page = NULL;

  int return_val;

  if (!pix_in)
    return TRUE;

  m_in = hocr_image_processing( pix_in, options,progress);
  if (!m_in)
    return TRUE;

  l_page = hocr_layout_analysis (m_in,
    font_spacing_code, paragraph_setup,
    slicing_threshold, slicing_width, line_leeway, dir_ltr, progress);

  if (!l_page)
  {
    ho_bitmap_free (m_in);
    return TRUE;
  }

  return_val = hocr_font_recognition (l_page, s_text_out,
    html, font_code, nikud, do_linguistics, progress);

  return return_val;
}

int
hocr_do_ocr (const ho_pixbuf * pix_in,
  ho_string * s_text_out,
  const unsigned char html, int font_code, const unsigned char do_linguistics,
  int *progress)
{
	HEBOCR_IMAGE_OPTIONS options;

	options.adaptive = 0;
	options.a_threshold = 0;
	options.no_auto_rotate = 0;
	options.no_auto_scale = 0;
	options.rotate = 0;
	options.scale = 0;
	options.threshold = 0;

	return hocr_do_ocr_fine( pix_in, s_text_out, &options,
     0, 0, 0, 0, 0, 0, html, font_code, 1, do_linguistics,
    progress);
}

const char *
hocr_get_build_string ()
{
  return BUILD;
}

const char *
hocr_get_version_string ()
{
  return VERSION;
}
